camviewer_ui_impl.py

```diff
@@ -1151,33 +1154,70 @@ def shutdown(self):
 
     def onfileSave(self):
         try:
-            fileName = QFileDialog.getSaveFileName(
-                self,
-                "Save Image...",
-                self.cwd,
-                "Images (*.npy *.jpg *.png *.bmp *.pgm *.tif)",
+            filename = QFileDialog.getSaveFileName(
+                parent=self,
+                caption="Save Image...",
+                directory=os.path.expanduser("~"),
+                filter="Images (*.npy *.jpg *.png *.bmp *.pgm *.tif)",
             )[0]
-            if fileName == "":
-                raise Exception("No File Name Specified")
-            if fileName.lower().endswith(".npy"):
-                np.save(fileName, self.image)
-                QMessageBox.information(
-                    self,
-                    "File Save Succeeded",
-                    "Image has been saved as a numpy file: %s" % (fileName),
+            if filename == "":
+                # The only way to get here is by clicking "X" or "Cancel"
+                return
+            file_ext = os.path.splitext(filename)[1]
+            if file_ext == ".npy":
+                # This either works or raises an OSError such as PermissionError
+                np.save(filename, self.image)
+                return self.show_file_success(filename=filename, file_ext=file_ext)
+            save_ok = self.ui.display_image.image.save(
+                filename, format=None, quality=-1
+            )
+            if save_ok:
+                # QImage.save returned True, so the save succeeded.
+                return self.show_file_success(filename=filename, file_ext=file_ext)
+            # QImage.save returned False, so the save failed.
+            # Check for obvious errors, then retry the save
+            # No write permissions?
+            directory = os.path.dirname(filename)
+            if not os.access(path=directory, mode=os.W_OK):
```

Comment:
If the file exists, should we check access permissions on the chosen file itself, rather than the directory?  Currently if I try to write to a file that exists but I have no permissions on, this will report an invalid image type...  (I wrote a file, write-protected it, then tried to save to that same file again)
<img width="497" alt="image" src="https://github.com/user-attachments/assets/13bb524a-6b4a-444f-ba84-52afea641cf5">

This also reveals a subtle bug where `image_types` has extensions without the `.`, but `file_ext` includes `.` (e.g. `.jpg`)

Reply:
Good catch. I originally was checking file permissions but that gave `False` if the file didn't exist. I'll check for existence and check the file or the folder as appropriate.

I'll clean up the user-facing reporting to make it more clear these are file extensions.

Reply:
I'm starting to regret not standing up unit tests for camviewer, as annoying as that would have been.